watertap/unit_models/ion_exchange_0D.py

```diff
@@ -0,0 +1,1346 @@
+        )
+
+        self.rinse_flow = Var(
+            initialize=0.001,
+            units=pyunits.m**3 / pyunits.s,
+            doc="Rinse step rate",
+        )
+
+        self.t_rinse = Var(
+            initialize=360, units=pyunits.s, bounds=(120, 1200), doc="Rinse time"
+        )
+
+        self.main_pump_power = Var(
+            initialize=4e-6, units=pyunits.kW, doc="Main pump power"
+        )
+
+        self.regen_pump_power = Var(
+            initialize=4e-6, units=pyunits.kW, doc="Regen pump power"
+        )
+
+        self.bw_pump_power = Var(
+            initialize=4e-6, units=pyunits.kW, doc="Backwash pump power"
+        )
+
+        self.rinse_pump_power = Var(
+            initialize=4e-6, units=pyunits.kW, doc="Rinse pump power"
+        )
+
+        self.pump_efficiency = Var(
+            initialize=0.8,
+            units=pyunits.dimensionless,
+            bounds=(0, 1),
+            doc="Pump efficiency",
+        )
+
+        tmp_dict = dict(**self.config.property_package_args)
+        tmp_dict["has_phase_equilibrium"] = False
+        tmp_dict["parameters"] = self.config.property_package
+        tmp_dict["defined_state"] = True  # inlet block is an inlet
+
+        self.properties_in = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,
+            doc="Material properties of inlet",
+            default=tmp_dict,
+        )
+
+        # Add outlet and waste block
+        tmp_dict["defined_state"] = False  # outlet and waste block is not an inlet
+        self.properties_out = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,
+            doc="Material properties of outlet",
+            default=tmp_dict,
+        )
+
+        self.properties_regen = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,
+            doc="Material properties of waste",
+            default=tmp_dict,
+        )
+
+        # Add ports - oftentimes users interact with these rather than the state blocks
+        self.add_port(name="inlet", block=self.properties_in)
+        self.add_port(name="outlet", block=self.properties_out)
+        self.add_port(name="waste", block=self.properties_regen)
+
+        # Add constraints
+        # =========== EQUILIBRIUM ===========
+
+        @self.Constraint(
+            ion_set,
+            doc="Separation factor calc",
+        )
+        def eq_sep_factor(b, j):
+            return b.separation_factor[j] == 1 / b.selectivity[j]
+
+        @self.Constraint(
+            ion_set,
+            doc="Langmuir isotherm",
+        )
+        def eq_langmuir(b, j):
+            if j == target_ion:
+                return b.separation_factor[j] == (
+                    b.c_norm[j] * (1 - b.resin_eq_capacity / b.resin_max_capacity)
+                ) / (b.resin_eq_capacity / b.resin_max_capacity * (1 - b.c_norm[j]))
+            else:
+                return Constraint.Skip
```

Comment:
This is a very odd form of Langmuir function. More commonly you see the Langmuir function take the following form:

`eq_cap = max_cap * (K * C) / (1 + K*C)`

Are we sure this is correct? Is this maybe not named correctly?

Reply:
Another thing to note here is that this constraint will probably cause convergence issues because you are dividing by a number that can become zero (i.e., the `1 - c_norm` part can become zero). 

You might want to consider reformatting this constraint to avoid the possibility of a divide by zero.

Reply:
Yes it is correct.